extensions/podman/packages/extension/src/podman-machine-stream.ts

```diff
@@ -0,0 +1,146 @@
+/**********************************************************************
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import * as fs from 'node:fs';
+
+import type {
+  Disposable,
+  Event,
+  ProviderConnectionShellAccess,
+  ProviderConnectionShellAccessData,
+  ProviderConnectionShellAccessError,
+  ProviderConnectionShellAccessSession,
+  ProviderConnectionShellDimensions,
+} from '@podman-desktop/api';
+import { EventEmitter } from '@podman-desktop/api';
+import type { ClientChannel } from 'ssh2';
+import { Client } from 'ssh2';
+
+import type { MachineInfo } from './extension';
+
+export class ProviderConnectionShellAccessImpl implements ProviderConnectionShellAccess, Disposable {
+  #host: string;
+  #port: number;
+  #username: string;
+  #privateKey: string;
+  #client: Client | undefined;
+  #stream: ClientChannel | undefined;
+  #connected: boolean = false;
+
+  constructor(private readonly podmanMachine: MachineInfo) {
+    this.#host = 'localhost';
+    this.#port = podmanMachine.port;
+    this.#username = podmanMachine.remoteUsername;
+    this.#privateKey = podmanMachine.identityPath;
+  }
+
+  dispose(): void {
+    this.close();
+  }
+
+  protected isConnected(): boolean {
+    return this.#connected;
+  }
+
+  onDataEmit = new EventEmitter<ProviderConnectionShellAccessData>();
+  onData: Event<ProviderConnectionShellAccessData> = this.onDataEmit.event;
+
+  onErrorEmit = new EventEmitter<ProviderConnectionShellAccessError>();
+  onError: Event<ProviderConnectionShellAccessError> = this.onErrorEmit.event;
+
+  onEndEmit = new EventEmitter<void>();
+  onEnd: Event<void> = this.onEndEmit.event;
+
+  write(data: string): void {
+    this.#stream?.write(data);
+  }
+
+  resize(dimensions: ProviderConnectionShellDimensions): void {
+    // rows and cols override width and height when rows and cols are non-zero.
+    this.#stream?.setWindow(dimensions.rows, dimensions.cols, 0, 0);
+  }
+
+  protected disposeListeners(): void {
+    this.onDataEmit.dispose();
+    this.onErrorEmit.dispose();
+    this.onEndEmit.dispose();
+  }
+
+  close(): void {
+    this.#connected = false;
+    this.#client?.end();
+    this.#client?.destroy();
+    this.disposeListeners();
```

Comment:
if there are two clients connected (that have called open), if the first one call close() it would close everything for both of them

this is why I think it's simpler/easier to not have the isConnected flag, and just always open when calling `open()` and closing when calling `close()`

Reply:
I would say that each connection has its session because each one of them will store the `ProviderConnectionShellAccessImpl` instance in its shellAccess. But the session of the given connection is opened only at the first call of "open". I would have to export the "close" function to the frontend to call it during onDestroy. This should result (+ some changes in this PR and another PR for exporting close function to backed) in making a new session on going to the "terminal tab"